_posts/2019-03-XX-5-terms-every-hypervisor-developer-should-know.md

```diff
@@ -0,0 +1,292 @@
+with a hypervisor. Hypercalls can be used for managing a virtual environment or
+to get some statistics from hypervisor - useful for debugging. Without these
+instructions, it is still possible to implement similar functionality, as we
+will see later in the series.
+
+## VMM
+
+Virtual machine monitor. It is a part of the software that handles all VM exits.
+Works in VMX root operation, it has full access to the hardware and almost full
+control of CPU. Why almost? It has some of the bits in control registers fixed
+to ensure proper execution. Trying to change CPU mode is one of the examples of
+such restrictions, so goodbye, real mode. Another thing that is limited in VMX
+root operation is that INIT signal is blocked - because of that it is impossible
+to reset other cores of CPU, which would effectively destroy everything that VMX
+gave.
+
+Other than those limitations, VMM can use everything that IA-32 has to offer.
+This includes interrupt and exception handlers (very useful, especially when
+passing instructions from VM to real hardware), virtual memory (might be used
+for easier mapping of VM memory, but it isn't common) and even going to ring 3
+(I don't think anyone would like to lower his/her privilege level here, but who
+knows?).
+
+It is VMM's task to isolate VMs from each other, if desired. VMM is the place
+where code and data for handlers is located, because of that it is crucial that
+VMM's memory is inaccessible from virtual machines. Such protection is possible
+with the help of EPT (extended page-table), but this mechanism is worthy of
+another post so I won't describe it further right now.
+
+Intel's *Software Developer’s Manual* calls this `host`, while most of the world
+leave this name for something else - I'll mention it later in this post. As
+with other multi-processor environments, we can develop symmetric and asymmetric
+VMMs. In this series, I will assume that a VMM on one core is a separate entity
+from another core, even on symmetrical systems. Hopefully, this will better show
+all nuances and possibilities of virtualization.
+
+A hypervisor is more than only its VMM part. It needs some initial setup, most
+likely a glue layer for the underlying system and sometimes checks for its
+starting arguments.
+
+## VM
+
+Virtual machine, also called guest software. It consists of OS and application
+software. Software running on VM usually is not aware of the fact that VMM
+controls hardware and that there may be other VMs running on the same platform,
+but it can be. Most of the times any application that can run on real hardware
+can also run inside a VM, depending on hypervisor implementation.
+
+As with VMM, VMs can also be symmetrical or asymmetrical when it comes to
+multi-processor execution. Again, I will restrict VM to one core. Here it makes
+even more sense because VMM can implement a scheduler and run multiple VMs on
+one physical core, in turn. For most people that were using only type-2
```

Comment:
@krystian-hebel maybe some basic explanation about types and examples of given types?

Reply:
done